### PR TITLE
Issue #427 suggestion

### PR DESCRIPTION
**Pull request by Dgmalthael**
<h1>Effects of Breaking News</h1>

Governmental and speculative news have significant impacts on the cryptocurrency market. Some unhealthy market behaviors can be attributed to news headlines [1]. Therefore, implementing this information into the market_health_reporter would result in more effective reasoning.

<h2>About the Implementation:</h2>

A function named sentiment_analysis is created in the tools/market_health_reporter/market_health_reporter.py file.

<h2>Usage:</h2>

A variable, efmm (news effect on market movement), is created and added to the prompt. The news titles will be checked and incorporated when an unhealthy market movement is found. Possible impacts of the news on the movement will be considered, enabling the market_health_reporter to provide better reasoning while generating outputs.

<h2>Summary:</h2>

The function scrapes news titles from Yahoo Finance and returns a list of dictionaries, which can also be represented as a data frame. The dictionaries consist of three main parts:

news_title: The data grabbed from Yahoo Finance.
label: The sentiment analysis of whether the news is positive, neutral, or negative.
score: The score of the sentiment label.

<h2>For example:</h2>

The following dictionary shows a positive label for the news title with a positivity score of 99 percent (adjustments may be needed for the percentage amount):

{
  'news_title': 'Bitcoin Futures Open Interest Hits 16-month High Amid Price Resilience',
  'label': 'positive',
  'score': 0.9981274008750916
}
<h2>Real-life Example:</h2>

News such as the Coinbase outage had a significant impact on Bitcoin, which led to unhealthy market behavior after Coinbase resumed operations [2].

<h2>Conclusion: </h2>
I believe, this is a necessary implementation for the market health reporter, however ,we need to work on the code to completely implement the sentimental analysis and we shall adjust some details such as score according to results. 

I am looking forward for your feedback 

<h2>Referance: </h2>
1.	https://www.sciencedirect.com/science/article/abs/pii/S0264999323002092#preview-section-cited-by
2.	https://cointelegraph.com/news/coinbase-currently-down-outage-ystem-wide

### CODE REVIEW

**Comment by evgenydmitriev** [comment]
@Dgmalthael, why did you choose to use a local sentiment model instead of relying on LLMs, like the rest of the Market Health Reporter?

**Comment by Dgmalthael** [comment]
@evgenydmitriev 
On the one hand,  while Python modules and general-purpose LLMs can indeed perform sentiment analysis, the customized model from Hugging Face was specifically trained for financial analysis. This customization is crucial, as financial texts have unique language patterns and terminologies that a generalized model might not handle with the same level of accuracy and nuance.

On the other hand, while ChatGPT and similar LLMs excel in a wide range of tasks, they have certain limitations when it comes to domain-specific analyses. For instance, generalized LLMs might not be as finely tuned to detect subtle sentiments or industry-specific jargon that are common in financial reports and market analysis. Using a dedicated financial sentiment model mitigates these risks and enhances the reliability of the insights derived.

To sum up, Using the local sentiment model was not mandatory however, I believe, it would perform better this way. 
I can try to customize the code "tools/python_modules"

**Comment by evgenydmitriev** [comment]
Hahaha are you using ChatGPT to generate reasons to not use it?

**Comment by Dgmalthael** [comment]
@evgenydmitriev

ChatGPT is a great tool to improve already existing ideas and create new ideas. 
Yes, I do use ChatGPT to improve my writing.

**I think, we are missing the point here, The main issue is improving the prompt and my idea is to add one extra variable in order to analyze the circulated news when there is unhealthy market behavior.** 

I can modify the code and tune it with your guidance  but **_what do you think about the idea and how can we improve it?_** 

I am looking forward for your feedback 😀

**Comment by evgenydmitriev** [comment]
Thank you for your contribution to the project and for taking the time to submit this pull request. Your efforts are appreciated, and I value your enthusiasm in suggesting improvements. I would like to provide some feedback to help align the proposed changes with the issue at hand, and hopefully, this will guide you toward a more effective solution.

#### Understanding the Issue at Hand

The original issue referenced aims at improving the prompt functionality. Specifically, the goal is to enhance how prompts are generated, structured, and potentially leveraged for better outcomes in various use cases. This involves focusing on the quality, clarity, and effectiveness of the prompts generated by the system.

#### Review of Proposed Changes

Your proposed code introduces Retrieval-Augmented Generation (RAG) functionality with a custom model, designed to be executed as a GitHub Action. While this is an innovative approach and RAG is indeed a powerful technique for combining the strengths of retrieval-based and generative models, it appears that this addition might not directly address the core of the original issue.

#### Alignment with the Original Issue

1. **Focus on Prompt Quality**:
    - The primary focus of the original issue is on improving prompt generation quality. While your approach brings in advanced functionality, it diverts from the main objective. Enhancing prompts requires refining the prompt generation process, possibly through better training data, fine-tuning, or incorporating user feedback mechanisms.

2. **Understanding Versus Generating**:
    - Leveraging Large Language Models (LLMs) is indeed beneficial. However, the key lies in using them to understand and improve existing prompts rather than generating complex functionalities that might not be necessary for the current scope. It's important to align with the issue's goal, which is prompt improvement, not the addition of new features.

#### Suggestions for Improvement

1. **Refinement of Existing Prompts**:
    - Consider focusing on how the LLMs can be used to evaluate and refine the prompts. This might involve techniques like prompt paraphrasing, syntactic and semantic analysis, and iterative testing to see what works best in various scenarios.

2. **User Feedback Integration**:
    - Incorporating a mechanism for users to provide feedback on the generated prompts could be immensely valuable. This feedback can be used to train and fine-tune the model, ensuring continuous improvement in prompt quality.

3. **Simplifying the Approach**:
    - While RAG is a powerful technique, it might be overcomplicating the solution for the current issue. Simplifying the approach to directly address prompt quality might be more effective. For instance, using transformers pre-trained on specific datasets relevant to the desired prompt types could yield better results.

Your efforts in adding advanced functionalities like RAG are commendable and showcase your technical skills and forward-thinking approach. However, aligning your contributions with the specific goals of the challenge is crucial. The best solutions are often the simplest ones that directly address the problem at hand.

To foster a more collaborative and focused development environment, here are a few steps you can consider:

1. **Clarify Objectives**:
    - Before diving into coding, ensure you have a clear understanding of the issue's objectives. Engaging in discussions with the challenge assignees or community can provide valuable insights.

2. **Incremental Improvements**:
    - Aim for incremental improvements that are directly aligned with the issue's goals. Small, focused changes can have a significant impact and are easier to review and integrate.

3. **Collaborative Feedback**:
    - Open the floor for collaborative feedback early in the development process. Share your initial ideas and get input from others to ensure you are on the right track.

In conclusion, while your proposed addition of RAG functionality is innovative, it may not be the best fit for resolving the issue of improving prompt quality. I encourage you to redirect your efforts towards refining the prompt generation process, leveraging LLMs for understanding and improving prompts rather than adding new functionalities. This focused approach will not only address the issue more effectively but also contribute to the overall enhancement of the Market Health Reporter.

Thank you once again for your contribution. Your dedication and hard work are highly valued, and I look forward to seeing your revised approach. Please do not hesitate to reach out if you have any questions or need further guidance.
